storage2/src/state.rs

```diff
@@ -0,0 +1,47 @@
+use std::collections::HashMap;
+
+use jmt::{
+    storage::{TreeReader, TreeWriter},
+    JellyfishMerkleTree,
+};
+
+mod transaction;
+use transaction::Transaction as StateTransaction;
+
+/// State is a lightweight copy-on-write fork of the chain state,
+/// implemented as a RYW cache over a pinned JMT version.
+pub(crate) struct State<'a, R: TreeReader + TreeWriter> {
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    jmt_version: jmt::Version,
+    jmt: &'a JellyfishMerkleTree<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> State<'a, R> {
```

Comment:
We can't have the `State` include a lifetime bound; that will make using it unworkable, since now every downstream consumer has to manage another object's lifetime and the relation between that object and the `State`.